main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -77,32 +77,350 @@ object Inliner {
+    *      the expression, recursively calling [[visitExp]] if it is a [[SubstRange.SuspendedExpr]].
+    *      This means that [[visitExp]] previously decided to unconditionally inline the let-binding
+    *      or decided to do copy-propagation of that binding (see below).
+    *   1. If `x'` is not in the expression substitution, it must be in the set of in-scope variable
+    *      definitions and considers it for inlining if it is pure.
+    *
+    * When [[visitExp]] encounters a let-binding `let sym = e1; e2` it considers five cases
+    * (note that it always refreshes `sym` to `sym'` as mentioned above):
+    *   1. If the binding is dead and pure, it drops the binding and returns `visitExp(e2)`.
+    *   1. If the binding is dead and impure, it rewrites the binding to a statement.
+    *   1. If the binding occurs once and is pure, it adds the unvisited `e1` to the substitution `subst`,
+    *      drops the binding and unconditionally inlines it at the occurrence of `sym`.
+    *   1. If the binding occurs more than once and is pure, it first visits `e1` and considers the following:
+    *      (a) If the visited `e1` is trivial, it removes the let-binding and unconditionally inlines
+    *      the visited `e1` at every occurrence of `sym`. This corresponds to copy-propagation.
+    *      (b) If the visited `e1` is not trivial, it keeps the let-binding, adds the visited `e1` to the set
+    *      of in-scope variable definitions and considers it for inlining at every occurrence.
+    *   1. If the binding occurs more than once and is impure, it keeps the let-binding and does not consider it for inlining.
+    */
+  private def visitExp(exp0: Expr, ctx0: LocalContext)(implicit sym0: Symbol.DefnSym, root: OccurrenceAst.Root, flix: Flix): Expr = exp0 match {
+    case Expr.Cst(cst, tpe, loc) =>
+      Expr.Cst(cst, tpe, loc)
+
+    case Expr.Var(sym, tpe, loc) =>
+      // Check for renamed local binder
+      ctx0.varSubst.get(sym) match {
+        case None => // Function parameter occurrence
+          Expr.Var(sym, tpe, loc)
+
+        case Some(freshVarSym) =>
+          ctx0.subst.get(freshVarSym) match {
+            // Case 1:
+            // The variable `sym` is in the substitution map. Replace `sym` with `e1`.
+            case Some(e1) =>
+              e1 match {
+                case SubstRange.DoneExpr(e) => // Reduced expression
+                  e
+
+                case SubstRange.SuspendedExpr(exp) => // Reduce suspended expr
+                  visitExp(exp, ctx0)
+              }
+
+            // Case 2:
+            // The variable `sym` is not in the substitution map, but is considered for inlining if it is pure.
+            case None =>
+              ctx0.inScopeVars.get(freshVarSym) match {
+                case Some(_) => Expr.Var(freshVarSym, tpe, loc)
+                case None => Expr.Var(freshVarSym, tpe, loc)
+              }
```

Comment:
This case is yet to be implemented, since this is were we will use heuristics. Currently we just do not inline anything at this point.

Reply:
Can we remove code that is not yet used?